### PR TITLE
Cache entry point lookups

### PR DESCRIPTION
**Pull request by danielhollas**
In #6091 we have sped up `aiida.orm` import significantly by delaying the call to `get_type_string_from_class`. However, we did not solve the underlying issue -- the reason why this call is so expensive is because it essentially involves a quadruply-nested for loop. 

Here I cache the results of `importlib_metadata.entry_points.select()` calls, which should get rid of the two inner loops coming from `importlib_metadata` code. This change shaves off around 25ms from `verdi` invocation, but should help throughout the code base. I am happy to do further benchmarking if needed (what should I benchmark?)

cc @sphuber

### CODE REVIEW

File: tests/plugins/test_entry_point.py [hunk 1]
```diff
@@ -72,10 +72,10 @@ def select(group, name):  # pylint: disable=unused-argument
 
 @pytest.mark.parametrize(
     'eps, name, exception', (
-        ((EP(name='ep', group='gr', value='x'),), None, None),
```
Comment (ltalirz) [inline tests/plugins/test_entry_point.py:75]: why is that change necessary?

File: aiida/plugins/entry_point.py [hunk 2]
```diff
@@ -30,11 +30,18 @@
 ENTRY_POINT_STRING_SEPARATOR = ':'
 
 
-@functools.lru_cache(maxsize=1)
+@functools.cache
 def eps():
     return _eps()
 
 
+@functools.lru_cache(maxsize=100)
```
Comment (ltalirz) [inline aiida/plugins/entry_point.py:38]: interesting, so do I understand correctly that `eps()` is cached, nothing is being read from disk here, and processing the few entry points in memory is still taking 25ms?

Can you please document this function to explain the quadruple loop problem here, and why this additional cache is necessary?

**Review by ltalirz — COMMENTED** [review]
Thanks a lot @danielhollas !

introducing caches is always increasing complexity,
but in this particular case I can't really come up with a scenario where this would hurt us (`eps()` is already being cached, so this should just be saving computational time)

so to me this looks fine

File: aiida/plugins/entry_point.py (hunk 2 above)
Comment (sphuber) [inline aiida/plugins/entry_point.py:38]: Are we worried about the size of the cache? I think the number of different calls to `eps_select` should be reasonable, not exceeding on the order of 100. So wouldn't we be better of with `lru_cache(maxsize=None)` i.e. `cache` which will be faster. Not sure how much faster it will be compared to an LRU cache of max 100 items. Might be negligible

File: aiida/manage/configuration/config.py [hunk 3]
```diff
@@ -28,7 +28,7 @@
 SCHEMA_FILE = 'config-v9.schema.json'
 
 
-@lru_cache(1)
```
Comment (sphuber) [inline aiida/manage/configuration/config.py:31]: I take it this is not strictly related, but I guess it couldn't hurt.

File: tests/plugins/test_entry_point.py (hunk 1 above)
Comment (sphuber) [reply tests/plugins/test_entry_point.py:75]: Together with the change below on line 95, the changes should have identical behavior. So I guess this is fine.

**Review by sphuber — COMMENTED** [review]
As @ltalirz I am bit hesitant whether this could introduce some subtle bugs related to cache invalidation, but couldn't come up with one for now. So think we can add this. Just have one comment

File: aiida/plugins/entry_point.py (hunk 2 above)
Comment (danielhollas) [reply aiida/plugins/entry_point.py:38]: I am actually worried. I wouldn't be surprised if the number of calls was bigger then 100, especially if plugins are installed, since in some functions we're essentially iterating over all existing entry points when looking for the entry point for a given class. I'll take a closer look and do some more benchmarking.

File: aiida/manage/configuration/config.py (hunk 3 above)
Comment (danielhollas) [reply aiida/manage/configuration/config.py:31]: Yes, not related. `@cache` was added in Python 3.9 so we can now use it since the 3.8 support was dropped.

File: aiida/plugins/entry_point.py (hunk 2 above)
Comment (danielhollas) [reply aiida/plugins/entry_point.py:38]: Yes, I'll write a comment. You can look at the cProfile info on the forum. https://aiida.discourse.group/t/why-is-aiidalab-base-widget-import-so-slow/32/16?u=danielhollas

Together with the fact that we were calling this during each Node class build (in a metaclass), this completely explained why the `aiida.orm` import was so slow. See #6091

Comment (ltalirz) [reply aiida/plugins/entry_point.py:38]: Are we doing this iteration or is it importlib?

Are we trying to find an entry point for a class without knowing the name of the entry point?

Comment (sphuber) [reply aiida/plugins/entry_point.py:38]: > since in some functions we're essentially iterating over all existing entry points when looking for the entry point for a given class. I'll take a closer look and do some more benchmarking.

These don't call this `eps_select` function though, do they? The cache here simply applies to the number of combinations of arguments with which it is called. Since it just has `group` and `name`, it should just be the list of all `(group, name)` tuples with which the function is called. This should, reasonably, not be much larger than the entry points that exist.

**Comment by danielhollas** [comment]
I looked a bit more carefully at the implementation of the `get_entry_point_from_class`.
In this function, we essentially need to iterate over all existing entry points (in my environment, there are 360 of them!). 
And we're doing this iteration very inefficiently, because we first grab the groups, and then filter entry points in each group. But this filtering is done using `EntryPoints.select()` which itself iterates over all the entry points.

Instead, we can simply iterate over all the entry points directly, see the last commit.

Here are some timings, using a non-existing class to measure the worst-case scenario:
(note that I have disabled cache for `get_entry_point_from_class` for these timings)

### Main branch

```python
In [1]: from aiida.plugins.entry_point import get_entry_point_from_class, eps
In [2]: p = eps() #warmup cache
In [3]: %timeit get_entry_point_from_class('invalid', 'invalid')
21.4 ms ± 108 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### Caching eps.select() on this branch

```python
In [1]: from aiida.plugins.entry_point import get_entry_point_from_class, eps
In [2]: p = eps(); get_entry_point_from_class('invalid', 'invalid') # warmup caches
In [12]: %timeit  get_entry_point_from_class('invalid', 'invalid')
329 µs ± 2.78 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

### Iterating over all entry points directly

```python
In [1]: from aiida.plugins.entry_point import get_entry_point_from_class, eps
In [2]: p = eps()
In [5]: %timeit get_entry_point_from_class('invalid', 'invalid')
288 µs ± 1.33 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

So the last approach clearly wins, and perhaps we could even remove the cache on top of this function.

I'll do a bit more investigation to see if it still makes sense to cache `eps.select() ` since it is called from other functions as well.

**Comment by sphuber** [comment]
> EDIT: The last commit does not work, need to investigate more..

The problem is that `get_entry_points` returns a list of `EntryPoint` whereas the `importlib_metadata.entry_points` function returns a list of strings. So you would have to load the entry point for each string in order to be able to get the module and class name. Not sure what the impact of having to load each entry point is going to be for the performance and whether the implementation will then still be faster than simply using `get_entry_points`.

**Comment by sphuber** [comment]
> EDIT: Updating 5.2.0 would require at the very least updating the EntryPointManager pytest fixture.

This should be straightforward. The `eps` has just changed from a nested mapping to a flat list, so instead of having `add_entry_point` treat it like a dict, we should simply append the entry point to the list. @danielhollas want me to take resolve this so we can wrap up this PR?

**Comment by danielhollas** [comment]
@sphuber So I should have clarified my hesitation here: I was not sure if you'd be okay with upgrading `importlib_metadata` because I wasn't sure if downstream packages might depend on this package as well. If that is not a concern I'll go ahead and finish this today or over the weekend.

> The eps has just changed from a nested mapping to a flat list, so instead of having add_entry_point treat it like a dict, we should simply append the entry point to the list. @danielhollas want me to take resolve this so we can wrap up this PR?

Indeed, but I stumbled at first because `EntryPoints` is in fact a subclass of `tuple` and not a `list` hence we'll probably need to recreate it instead of modifying it directly.

**Comment by sphuber** [comment]
No, I think it is fine for use to update the minimum requirement of `importlib_metadat`. Please feel free to go ahead with this

File: aiida/manage/tests/pytest_fixtures.py [hunk 4]
```diff
@@ -835,31 +840,23 @@ def remove(
...
-
-    .. note:: This fixture does not use the ``suppress_deprecations`` decorator on purpose, but instead adds it manually
-        inside the fixture's body. The reason is that otherwise all deprecations would be suppressed for the entire
-        scope of the fixture, including those raised by the code run in the test using the fixture, which is not
-        desirable.
-
+    This fixture monkey patches the entry point caches returned by the :func:`aiida.plugins.entry_point.eps` and
+    :func:`aiida.plugins.entry_point.eps_select` methods to a class methods of the EntryPointManager
+    so that we can dynamically add / remove entry points.
+    Note that we do not need a deepcopy here is eps() returns an immutable EntryPoints tuple type.
```
Comment (sphuber) [inline aiida/manage/tests/pytest_fixtures.py:857]:
```suggestion
    Note that we do not need a deepcopy here as ``eps()`` returns an immutable ``EntryPoints`` tuple type.
```

File: aiida/manage/tests/pytest_fixtures.py [hunk 5]
```diff
@@ -835,31 +840,23 @@ def remove(
...
-    method and then monkey patches that function to return the deepcopy. This ensures that the changes on the entry
-    point cache performed during the test through the manager are undone at the end of the function scope.
-
-    .. note:: This fixture does not use the ``suppress_deprecations`` decorator on purpose, but instead adds it manually
-        inside the fixture's body. The reason is that otherwise all deprecations would be suppressed for the entire
-        scope of the fixture, including those raised by the code run in the test using the fixture, which is not
-        desirable.
-
+    This fixture monkey patches the entry point caches returned by the :func:`aiida.plugins.entry_point.eps` and
+    :func:`aiida.plugins.entry_point.eps_select` methods to a class methods of the EntryPointManager
```
Comment (sphuber) [inline aiida/manage/tests/pytest_fixtures.py:855]:
```suggestion
    :func:`aiida.plugins.entry_point.eps_select` functions to a class methods of the ``EntryPointManager``
```

File: aiida/plugins/entry_point.py [hunk 6]
```diff
@@ -30,11 +30,18 @@
...
 
-@functools.lru_cache(maxsize=1)
-def eps():
+@functools.cache
+def eps() -> EntryPoints:
     return _eps()
 
 
+@functools.lru_cache(maxsize=100)
+def eps_select(group, name=None) -> EntryPoints:
```
Comment (sphuber) [inline aiida/plugins/entry_point.py:39]: Can you please add a docstring here and add a note that you are providing this over `eps.select` in order to be able to cache the calls

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @danielhollas . Implementation looks good, just some minor docstring requests and then we can merge

**Comment by danielhollas** [comment]
> Just checking whether this is then indeed still necessary or whether we can reduce the caching

So some timings: tl;dr: entry_points.select() is a fairly expensive operation and we should probably cache it since aiida uses entry points extensively:

```python
# Without cache
In [7]: %timeit eps.select(group='invalid', name='invalid')
363 µs ± 4.13 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
In [8]: %timeit eps.select(group='console_scripts', name='verdi')
376 µs ± 14.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

# with cache
%timeit eps_select(group='console_scripts', name='verdi')
201 ns ± 0.886 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

The `get_entry_point_from_class` now takes similer in the order of hundreds of microseconds, so at least order of magnitude faster, but still 3 orders of magnitude slower than a cache, so removing that cache might have some unintended consequences in some cases.
We can nonetheless achieve further speed up by pre-sorting the `EntryPoints` by the group name so that `aiida.` groups come first. This can give large speedups in some cases.

```python
# Best case scenario, first item in sorted `EntryPoints`
In [64]: %timeit get_entry_point_from_class_uncached('aiida.calculations.arithmetic.add', 'ArithmeticAddCalculation')
2.4 µs ± 11.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

# Last item in aiida groups
In [65]: %timeit get_entry_point_from_class_uncached('aiida.workflows.arithmetic.add_multiply', 'add_multiply')
83.4 µs ± 362 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

# Iterating through all entry points (243 in this case)
%timeit get_entry_point_from_class_uncached('invalid', 'invalid')
231 µs ± 416 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each) 
```

The pre-sorting itself takes ~30µs so its worth it in most cases
```python
 %timeit EntryPoints(sorted(eps_unsorted, key=lambda x: x.group))
37.2 µs ± 113 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

File: environment.yml [hunk 7]
```diff
@@ -20,7 +20,7 @@ dependencies:
 - jinja2~=3.0
 - jsonschema~=3.0
 - kiwipy[rmq]~=0.7.7
-- importlib-metadata~=4.13
+- importlib-metadata~=6.0
```
Comment (danielhollas) [inline environment.yml:23]: @sphuber I've bumped this all the way to 6.x and test still pass. WDYT?

File: aiida/orm/autogroup.py [hunk 8]
```diff
@@ -86,7 +86,7 @@ def validate(strings: list[str] | None):
         """Validate the list of strings passed to set_include and set_exclude."""
         if strings is None:
             return
-        valid_prefixes = set(['aiida.node', 'aiida.calculations', 'aiida.workflows', 'aiida.data'])
```
Comment (danielhollas) [inline aiida/orm/autogroup.py:89]: Just an unrelated microoptimization

**Review by danielhollas — COMMENTED** [review]
@sphuber thank you for the review. I've modified the docstrings, and added the EntryPoints pre-sorting, as described above.

File: environment.yml (hunk 7 above)
Comment (sphuber) [reply environment.yml:23]: It should be compatible with Python 3.8 and up and not have any other significant dependency restrictions, so I think this is fine: https://github.com/python/importlib_metadata/blob/main/setup.cfg

**Review by sphuber — APPROVED** [review]

**Comment by ltalirz** [comment]
Thanks @danielhollas!
